src/catalogue.js

```diff
@@ -0,0 +1,112 @@
+import { first, pick, get } from 'lodash';
+import * as xhr from './xhr';
+import { mdToExecutionConfiguration } from './execution';
+
+const BASIC_PROPS = ['title', 'summary', 'identifier', 'uri'];
+const REQUEST_DEFAULTS = {
+    types: ['attribute', 'metric', 'fact'],
+    paging: {
+        offset: 0
+    }
+};
+
+function pickBasicProps(item) {
+    return pick(item.meta, BASIC_PROPS);
+}
+
+function deserializeFact(fact) {
+    return pickBasicProps(fact);
+}
+
+function deserializeMetric(metric) {
+    const basicProps = pickBasicProps(metric);
+    const metricProps = pick(metric.content, ['expression', 'format']);
+
+    return { ...basicProps, ...metricProps };
+}
+
+function deserializeAttribute(attribute) {
+    const basicProps = pickBasicProps(attribute);
+    const displayForm = first(attribute.content.displayForms);
+    const attributeProps = {
+        dimensionUri: get(attribute, ['content', 'dimension']),
+        granularity: get(attribute, ['meta', 'identifier']),
+        elementsUri: get(displayForm, ['links', 'elements']),
+        dfIdentifier: get(displayForm, ['meta', 'identifier']),
+        dfUri: get(displayForm, ['meta', 'uri']),
+        dateType: get(attribute, ['content', 'type'])
+    };
+
+    return { ...basicProps, ...attributeProps };
+}
+
+const DESERIALIZERS = {
+    fact: deserializeFact,
+    metric: deserializeMetric,
+    attribute: deserializeAttribute
+};
+
+function deserializeItem(item) {
+    const type = first(Object.keys(item));
+
+    if (!DESERIALIZERS.hasOwnProperty(type)) {
+        throw Error(`Can't deserialize item of type: ${type}`);
+    }
+
+    const props = DESERIALIZERS[type](item[type]);
+
+    return {
+        ...props,
+        type,
+        id: props.identifier,
+        isAvailable: true
+    };
+}
+
+function deserializeItems(items) {
+    return items.map(deserializeItem);
+}
+
+function bucketItemsToExecConfig(bucketItems) {
+    const executionConfig = mdToExecutionConfiguration(bucketItems);
+    const definitions = get(executionConfig, 'execution.definitions');
+
+    return get(executionConfig, 'execution.columns').map((column) => {
+        const definition = definitions.find(({ metricDefinition }) => get(metricDefinition, 'identifier') === column);
+        const maql = get(definition, 'metricDefinition.expression');
```

Comment:
How's it going work for one metric with multiple metric definitions?

For example fact based metric with contribution?